src/validation.cpp

```diff
@@ -1858,12 +1858,58 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
     // before the first had been spent.  Since those coinbases are sufficiently buried its no longer possible to create further
     // duplicate transactions descending from the known pairs either.
     // If we're on the known chain at height greater than where BIP34 activated, we can save the db accesses needed for the BIP30 check.
+
+    // The exception to this logic is if there are coinbases from blocks before
```

Comment:
Rather than start your new comment explaining where the previous comment is incorrect, I think it makes sense to remove/reword the paragraph above.

I also think it would make the whole comment clearer if you explicitly introduced terminology at the top of the comment. Something like:

```
- block height: the height of the block in the blockchain
- coinbase height: the height indicated in the coinbase transaction's scriptSig under BIP34 rules.

All blocks with block height >= 227,931 must have coinbase height equal to block height. This is a consensus rule.
```
rather than sprinkling the rest of the comment with multiple terms for the same concept: `a scriptSig with a valid-looking BIP34 height...`, `scriptSig indicating...`, `indicated block height`, `coinbase with indicated height`.

Reply:
I think it is useful to understand the basic logic that BIP 34 implies BIP 30 is met and doesn't need to be checked.  But then understand why it's not quite that simple.  It also matches the historical evolution of the code.  So I left this comment as correction to the previous one.